src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -7792,6 +7792,231 @@ struct TransposeReduceSimplify : public OpRewritePattern<stablehlo::ReduceOp> {
+      return failure();
+
+    Value condValue = nullptr;
+    bool positive = true;
+    auto lhsCompareOp = compareOp.getLhs();
+    auto rhsCompareOp = compareOp.getRhs();
+    if (matchPattern(lhsCompareOp, m_AnyZeroFloat()) ||
+        matchPattern(lhsCompareOp, m_Zero())) {
+      condValue = compareOp.getRhs();
+      positive = compareOp.getComparisonDirection() ==
+                     stablehlo::ComparisonDirection::GT ||
+                 compareOp.getComparisonDirection() ==
+                     stablehlo::ComparisonDirection::GE;
+    } else if (matchPattern(rhsCompareOp, m_AnyZeroFloat()) ||
+               matchPattern(rhsCompareOp, m_Zero())) {
+      condValue = compareOp.getLhs();
+      positive = compareOp.getComparisonDirection() ==
+                     stablehlo::ComparisonDirection::LT ||
+                 compareOp.getComparisonDirection() ==
+                     stablehlo::ComparisonDirection::LE;
+    } else {
+      return failure();
+    }
+
+    auto newOp = rewriter.create<stablehlo::SignOp>(op.getLoc(), condValue);
+    if (positive) { // cond > or >= 0
+      if (lhspositive) {
+        rewriter.replaceOpWithNewOp<stablehlo::MulOp>(op, newOp, rhs);
+      } else {
+        auto negRhs = rewriter.create<stablehlo::NegOp>(op.getLoc(), rhs);
+        rewriter.replaceOpWithNewOp<stablehlo::MulOp>(op, newOp, negRhs);
+      }
+    } else { // cond < or <= 0
+      if (lhspositive) {
+        auto negRhs = rewriter.create<stablehlo::NegOp>(op.getLoc(), rhs);
+        rewriter.replaceOpWithNewOp<stablehlo::MulOp>(op, newOp, negRhs);
+      } else {
+        rewriter.replaceOpWithNewOp<stablehlo::MulOp>(op, newOp, rhs);
+      }
+    }
+
+    return success();
+  }
+};
+
+// (mul (sign x) (abs x)) -> x
+// (mul (abs x) (sign x)) -> x
+struct SignAbsSimplify : public OpRewritePattern<stablehlo::MulOp> {
+  using OpRewritePattern<stablehlo::MulOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(stablehlo::MulOp op,
+                                PatternRewriter &rewriter) const override {
+    auto lhs = op.getOperand(0);
+    auto rhs = op.getOperand(1);
+
+    auto lhsSignOp = lhs.getDefiningOp<stablehlo::SignOp>();
+    if (lhsSignOp) {
+      auto rhsAbsOp = rhs.getDefiningOp<stablehlo::AbsOp>();
+      if (!rhsAbsOp)
+        return failure();
+
+      if (lhsSignOp.getOperand() != rhsAbsOp.getOperand())
+        return failure();
+
+      rewriter.replaceOp(op, lhsSignOp.getOperand());
+      return success();
+    }
+
+    auto rhsSignOp = rhs.getDefiningOp<stablehlo::SignOp>();
+    if (rhsSignOp) {
+      auto lhsAbsOp = lhs.getDefiningOp<stablehlo::AbsOp>();
+      if (!lhsAbsOp)
+        return failure();
+
+      if (rhsSignOp.getOperand() != lhsAbsOp.getOperand())
+        return failure();
+
+      rewriter.replaceOp(op, rhsSignOp.getOperand());
+      return success();
+    }
+
+    return failure();
+  }
+};
+
+// (mul (neg x) (neg y)) -> (mul x y)
+// (mul (neg x) y) -> (neg (mul x y))
```

Comment:
top one is always good, the one negation variations we should separate since there's a separate question of whether we want to propagate them up or down (e.g. if we had mul (neg x), constant) we'd want to do mul x (-constant)